File: sdk/typescript/src/api/keys.ts
import type { HttpClient } from "../http.js";
import type {
  KeyBundle,
  KeyHealth,
  PreKeysRequest,
  SignedPreKeyRequest,
} from "../types/index.js";

export class KeysApi {
  constructor(private readonly http: HttpClient) {}

  getBundle(agentId: string): Promise<KeyBundle> {
    return this.http.get<KeyBundle>(
      `/keys/${encodeURIComponent(agentId)}/bundle`,
    );
  }

  health(agentId: string): Promise<KeyHealth> {
    return this.http.getDirectoryAuthAs<KeyHealth>(
      `/keys/${encodeURIComponent(agentId)}/health`,
      agentId,
    );
  }

  uploadPreKeys(agentId: string, request: PreKeysRequest): Promise<void> {
    return this.http.putDirectoryAuthAs<void>(
      `/keys/${encodeURIComponent(agentId)}/prekeys`,
      agentId,
      request,
    );
  }

  rotateSignedPreKey(
    agentId: string,
    request: SignedPreKeyRequest,
  ): Promise<void> {
    return this.http.putDirectoryAuthAs<void>(
      `/keys/${encodeURIComponent(agentId)}/signed-prekey`,
      agentId,
      request,
    );
  }
}

File: sdk/python/src/tinyplace/api/keys.py
from __future__ import annotations

from ..http import HttpClient, encode
from ..types import Json, JsonDict


class KeysApi:
    def __init__(self, http: HttpClient) -> None:
        self._http = http

    async def get_bundle(self, agent_id: str) -> Json:
        return await self._http.get(f"/keys/{encode(agent_id)}/bundle")

    async def health(self, agent_id: str) -> Json:
        return await self._http.get_directory_auth_as(
            f"/keys/{encode(agent_id)}/health",
            agent_id,
        )

    async def upload_pre_keys(self, agent_id: str, request: JsonDict) -> None:
        await self._http.put_directory_auth_as(
            f"/keys/{encode(agent_id)}/prekeys",
            agent_id,
            request,
        )

    async def rotate_signed_pre_key(self, agent_id: str, request: JsonDict) -> None:
        await self._http.put_directory_auth_as(
            f"/keys/{encode(agent_id)}/signed-prekey",
            agent_id,
            request,
        )
